syntax = "proto2";

package dali_proto;

// Stores a single argument
message Argument {
  required string name = 1;
  optional string type = 2;
  repeated float floats = 3;
  repeated int64 ints = 4;
  repeated string strings = 5;
  repeated bool bools = 6;
  repeated Argument extra_args = 7;

  // vector storage
  required bool is_vector = 8 [default = false];
}

message InputOutput {
  required string name = 1;
  required string device = 2;
  required bool is_argument_input = 3 [default = false];
  optional string arg_name = 4;
}

// stores info about a single operator
message OpDef {
  required string name = 1;
  // list of inputs & outputs
  repeated InputOutput input = 2;
  repeated InputOutput output = 3;
  // Arguments
  repeated Argument args = 4;
  // instance name
  optional string inst_name = 5;
}

// Stores pipeline options and a list of operators
message PipelineDef {
  optional int64 num_threads = 1 [default = 4];
  required int32 batch_size = 2;
  optional bool use_async = 3 [default = false];
  optional bool pipeline = 4 [default = false];

  // Store all container operators
  repeated OpDef op = 5;

  // Explicitly store all external inputs
  repeated string external_inputs = 6;

  // Store all registered outputs
  repeated InputOutput pipe_outputs = 7;
  optional int32 device_id = 8 [default = 0];
}
